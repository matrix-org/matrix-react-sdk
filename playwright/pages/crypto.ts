/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { APIRequestContext, Page, expect } from "@playwright/test";

import { HomeserverInstance } from "../plugins/homeserver";

export class Crypto {
    public constructor(
        private page: Page,
        private homeserver: HomeserverInstance,
        private request: APIRequestContext,
    ) {}

    /**
     * Check that the user has published cross-signing keys, and that the user's device has been cross-signed.
     */
    public async assertDeviceIsCrossSigned(): Promise<void> {
        const { userId, deviceId, accessToken } = await this.page.evaluate(() => ({
            userId: window.mxMatrixClientPeg.get().getUserId(),
            deviceId: window.mxMatrixClientPeg.get().getDeviceId(),
            accessToken: window.mxMatrixClientPeg.get().getAccessToken(),
        }));

        const res = await this.request.post(`${this.homeserver.config.baseUrl}/_matrix/client/v3/keys/query`, {
            headers: { Authorization: `Bearer ${accessToken}` },
            data: { device_keys: { [userId]: [] } },
        });
        const json = await res.json();

        // there should be three cross-signing keys
        expect(json.master_keys[userId]).toHaveProperty("keys");
        expect(json.self_signing_keys[userId]).toHaveProperty("keys");
        expect(json.user_signing_keys[userId]).toHaveProperty("keys");

        // and the device should be signed by the self-signing key
        const selfSigningKeyId = Object.keys(json.self_signing_keys[userId].keys)[0];

        expect(json.device_keys[userId][deviceId]).toBeDefined();

        const myDeviceSignatures = json.device_keys[userId][deviceId].signatures[userId];
        expect(myDeviceSignatures[selfSigningKeyId]).toBeDefined();
    }
}
