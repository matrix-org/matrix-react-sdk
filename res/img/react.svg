<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
