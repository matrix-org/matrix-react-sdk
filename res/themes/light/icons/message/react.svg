<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>