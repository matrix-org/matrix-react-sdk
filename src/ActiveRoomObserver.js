/*
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import RoomViewStore from './stores/RoomViewStore';

/**
 * Consumes changes from the RoomViewStore and notifies specific things
 * about when the active room changes. Unlike listening for RoomViewStore
 * changes, you can subscribe to only changes relevant to a particular
 * room.
 *
 * TODO: If we introduce an observer for something else, factor out
 * the adding / removing of listeners & emitting into a common class.
 */
class ActiveRoomObserver {
    constructor() {
        this._listeners = {};

        this._activeRoomId = RoomViewStore.getRoomId();
        // TODO: We could self-destruct when the last listener goes away, or at least
        // stop listening.
        this._roomStoreToken = RoomViewStore.addListener(this._onRoomViewStoreUpdate.bind(this));
    }

    addListener(roomId, listener) {
        if (!this._listeners[roomId]) this._listeners[roomId] = [];
        this._listeners[roomId].push(listener);
    }

    removeListener(roomId, listener) {
        if (this._listeners[roomId]) {
            const i = this._listeners[roomId].indexOf(listener);
            if (i > -1) {
                this._listeners[roomId].splice(i, 1);
            }
        } else {
            console.warn("Unregistering unrecognised listener (roomId=" + roomId + ")");
        }
    }

    _emit(roomId) {
        if (!this._listeners[roomId]) return;

        for (const l of this._listeners[roomId]) {
            l.call();
        }
    }

    _onRoomViewStoreUpdate() {
        // emit for the old room ID
        if (this._activeRoomId) this._emit(this._activeRoomId);

        // update our cache
        this._activeRoomId = RoomViewStore.getRoomId();

        // and emit for the new one
        if (this._activeRoomId) this._emit(this._activeRoomId);
    }
}

if (global.mx_ActiveRoomObserver === undefined) {
    global.mx_ActiveRoomObserver = new ActiveRoomObserver();
}
export default global.mx_ActiveRoomObserver;
