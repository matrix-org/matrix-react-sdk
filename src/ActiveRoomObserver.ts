/*
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import RoomViewStore from './stores/RoomViewStore';

type Listener = (isActive: boolean) => void;

/**
 * Consumes changes from the RoomViewStore and notifies specific things
 * about when the active room changes. Unlike listening for RoomViewStore
 * changes, you can subscribe to only changes relevant to a particular
 * room.
 *
 * TODO: If we introduce an observer for something else, factor out
 * the adding / removing of listeners & emitting into a common class.
 */
export class ActiveRoomObserver {
    private listeners: {[key: string]: Listener[]} = {};
    private _activeRoomId = RoomViewStore.getRoomId();
    private readonly roomStoreToken: string;

    constructor() {
        // TODO: We could self-destruct when the last listener goes away, or at least stop listening.
        this.roomStoreToken = RoomViewStore.addListener(this.onRoomViewStoreUpdate);
    }

    public get activeRoomId(): string {
        return this._activeRoomId;
    }

    public addListener(roomId, listener) {
        if (!this.listeners[roomId]) this.listeners[roomId] = [];
        this.listeners[roomId].push(listener);
    }

    public removeListener(roomId, listener) {
        if (this.listeners[roomId]) {
            const i = this.listeners[roomId].indexOf(listener);
            if (i > -1) {
                this.listeners[roomId].splice(i, 1);
            }
        } else {
            console.warn("Unregistering unrecognised listener (roomId=" + roomId + ")");
        }
    }

    private emit(roomId, isActive: boolean) {
        if (!this.listeners[roomId]) return;

        for (const l of this.listeners[roomId]) {
            l.call(null, isActive);
        }
    }

    private onRoomViewStoreUpdate = () => {
        // emit for the old room ID
        if (this._activeRoomId) this.emit(this._activeRoomId, false);

        // update our cache
        this._activeRoomId = RoomViewStore.getRoomId();

        // and emit for the new one
        if (this._activeRoomId) this.emit(this._activeRoomId, true);
    };
}

if (window.mxActiveRoomObserver === undefined) {
    window.mxActiveRoomObserver = new ActiveRoomObserver();
}
export default window.mxActiveRoomObserver;
