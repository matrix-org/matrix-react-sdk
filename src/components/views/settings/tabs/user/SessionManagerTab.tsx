/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { useState } from 'react';

import { _t } from "../../../../../languageHandler";
import { useOwnDevices } from '../../devices/useOwnDevices';
import SettingsSubsection from '../../shared/SettingsSubsection';
import FilteredDeviceList from '../../devices/FilteredDeviceList';
import CurrentDeviceSection from '../../devices/CurrentDeviceSection';
import SecurityRecommendations from '../../devices/SecurityRecommendations';
import { DeviceSecurityVariation, DeviceWithVerification } from '../../devices/types';
import SettingsTab from '../SettingsTab';

const SessionManagerTab: React.FC = () => {
    const { devices, currentDeviceId, isLoading } = useOwnDevices();
    const [filter, setFilter] = useState<DeviceSecurityVariation>();
    const [expandedDeviceIds, setExpandedDeviceIds] = useState([]);

    const onDeviceExpandToggle = (deviceId: DeviceWithVerification['device_id']): void => {
        if (expandedDeviceIds.includes(deviceId)) {
            setExpandedDeviceIds(expandedDeviceIds.filter(id => id !== deviceId));
        } else {
            setExpandedDeviceIds([...expandedDeviceIds, deviceId]);
        }
    };

    const { [currentDeviceId]: currentDevice, ...otherDevices } = devices;
    const shouldShowOtherSessions = Object.keys(otherDevices).length > 0;

    return <SettingsTab heading={_t('Sessions')}>
        <SecurityRecommendations devices={devices} />
        <CurrentDeviceSection
            device={currentDevice}
            isLoading={isLoading}
        />
        {
            shouldShowOtherSessions &&
            <SettingsSubsection
                heading={_t('Other sessions')}
                description={_t(
                    `For best security, verify your sessions and sign out ` +
                    `from any session that you don't recognize or use anymore.`,
                )}
                data-testid='other-sessions-section'
            >
                <FilteredDeviceList
                    devices={otherDevices}
                    filter={filter}
                    expandedDeviceIds={expandedDeviceIds}
                    onFilterChange={setFilter}
                    onDeviceExpandToggle={onDeviceExpandToggle}
                />
            </SettingsSubsection>
        }
    </SettingsTab>;
};

export default SessionManagerTab;
