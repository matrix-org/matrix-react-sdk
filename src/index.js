/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
