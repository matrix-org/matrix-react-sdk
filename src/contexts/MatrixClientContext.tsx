/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, {
    ComponentClass,
    createContext,
    forwardRef,
    PropsWithoutRef,
    ForwardRefExoticComponent,
    useContext,
    RefAttributes,
} from "react";
import { MatrixClient } from "matrix-js-sdk/src/matrix";

// This context is available to components under LoggedInView,
// the context must not be used by components outside a MatrixClientContext tree.
// This assertion allows us to make the type not nullable.
const MatrixClientContext = createContext<MatrixClient>(null as any);
MatrixClientContext.displayName = "MatrixClientContext";
export default MatrixClientContext;

export interface MatrixClientProps {
    mxClient: MatrixClient;
}

export function useMatrixClientContext(): MatrixClient {
    return useContext(MatrixClientContext);
}

const matrixHOC = <ComposedComponentProps extends {}>(
    ComposedComponent: ComponentClass<ComposedComponentProps>,
): ForwardRefExoticComponent<
    PropsWithoutRef<Omit<ComposedComponentProps, "mxClient">> &
        RefAttributes<InstanceType<ComponentClass<ComposedComponentProps>>>
> => {
    type ComposedComponentInstance = InstanceType<typeof ComposedComponent>;

    // eslint-disable-next-line react-hooks/rules-of-hooks

    const TypedComponent = ComposedComponent;

    return forwardRef<ComposedComponentInstance, Omit<ComposedComponentProps, "mxClient">>((props, ref) => {
        const client = useContext(MatrixClientContext);

        // @ts-ignore
        return <TypedComponent ref={ref} {...props} mxClient={client} />;
    });
};
export const withMatrixClientHOC = matrixHOC;
