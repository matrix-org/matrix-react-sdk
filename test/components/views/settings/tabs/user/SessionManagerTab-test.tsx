/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { DeviceInfo } from 'matrix-js-sdk/src/crypto/deviceinfo';
import { logger } from 'matrix-js-sdk/src/logger';
import { DeviceTrustLevel } from 'matrix-js-sdk/src/crypto/CrossSigning';

import SessionManagerTab from '../../../../../../src/components/views/settings/tabs/user/SessionManagerTab';
import MatrixClientContext from '../../../../../../src/contexts/MatrixClientContext';
import {
    flushPromisesWithFakeTimers,
    getMockClientWithEventEmitter,
    mockClientMethodsUser,
} from '../../../../../test-utils';

jest.useFakeTimers();

describe('<SessionManagerTab />', () => {
    const aliceId = '@alice:server.org';
    const deviceId = 'alices_device';

    const alicesDevice = {
        device_id: deviceId,
    };
    const alicesMobileDevice = {
        device_id: 'alices_mobile_device',
        last_seen_ts: Date.now(),
    };

    const alicesOlderMobileDevice = {
        device_id: 'alices_older_mobile_device',
        last_seen_ts: Date.now() - 600000,
    };

    const mockCrossSigningInfo = {
        checkDeviceTrust: jest.fn(),
    };
    const mockClient = getMockClientWithEventEmitter({
        ...mockClientMethodsUser(aliceId),
        getStoredCrossSigningForUser: jest.fn().mockReturnValue(mockCrossSigningInfo),
        getDevices: jest.fn(),
        getStoredDevice: jest.fn(),
        getDeviceId: jest.fn().mockReturnValue(deviceId),
    });

    const defaultProps = {};
    const getComponent = (props = {}): React.ReactElement =>
        (
            <MatrixClientContext.Provider value={mockClient}>
                <SessionManagerTab {...defaultProps} {...props} />
            </MatrixClientContext.Provider>
        );

    beforeEach(() => {
        jest.clearAllMocks();
        jest.spyOn(logger, 'error').mockRestore();
        mockClient.getDevices.mockResolvedValue({ devices: [] });
        mockClient.getStoredDevice.mockImplementation((_userId, id) => {
            const device = [alicesDevice, alicesMobileDevice].find(device => device.device_id === id);
            return device ? new DeviceInfo(device.device_id) : null;
        });
        mockCrossSigningInfo.checkDeviceTrust
            .mockReset()
            .mockReturnValue(new DeviceTrustLevel(false, false, false, false));
    });

    it('renders spinner while devices load', () => {
        const { container } = render(getComponent());
        expect(container.getElementsByClassName('mx_Spinner').length).toBeTruthy();
    });

    it('removes spinner when device fetch fails', async () => {
        mockClient.getDevices.mockRejectedValue({ httpStatus: 404 });
        const { container } = render(getComponent());
        expect(mockClient.getDevices).toHaveBeenCalled();

        await act(async () => {
            await flushPromisesWithFakeTimers();
        });
        expect(container.getElementsByClassName('mx_Spinner').length).toBeFalsy();
    });

    it('removes spinner when device fetch fails', async () => {
        // eat the expected error log
        jest.spyOn(logger, 'error').mockImplementation(() => {});
        mockClient.getDevices.mockRejectedValue({ httpStatus: 404 });
        const { container } = render(getComponent());

        await act(async () => {
            await flushPromisesWithFakeTimers();
        });
        expect(container.getElementsByClassName('mx_Spinner').length).toBeFalsy();
    });

    it('does not fail when checking device verification fails', async () => {
        const logSpy = jest.spyOn(logger, 'error').mockImplementation(() => {});
        mockClient.getDevices.mockResolvedValue({ devices: [alicesDevice, alicesMobileDevice] });
        const noCryptoError = new Error("End-to-end encryption disabled");
        mockClient.getStoredDevice.mockImplementation(() => { throw noCryptoError; });
        render(getComponent());

        await act(async () => {
            await flushPromisesWithFakeTimers();
        });

        // called for each device despite error
        expect(mockClient.getStoredDevice).toHaveBeenCalledWith(aliceId, alicesDevice.device_id);
        expect(mockClient.getStoredDevice).toHaveBeenCalledWith(aliceId, alicesMobileDevice.device_id);
        expect(logSpy).toHaveBeenCalledWith('Error getting device cross-signing info', noCryptoError);
    });

    it('sets device verification status correctly', async () => {
        mockClient.getDevices.mockResolvedValue({ devices: [alicesDevice, alicesMobileDevice] });
        mockCrossSigningInfo.checkDeviceTrust
            // alices device is trusted
            .mockReturnValueOnce(new DeviceTrustLevel(true, true, false, false))
            // alices mobile device is not
            .mockReturnValueOnce(new DeviceTrustLevel(false, false, false, false));

        const { getByTestId } = render(getComponent());

        await act(async () => {
            await flushPromisesWithFakeTimers();
        });

        expect(mockCrossSigningInfo.checkDeviceTrust).toHaveBeenCalledTimes(2);
        expect(getByTestId(`device-tile-${alicesDevice.device_id}`)).toMatchSnapshot();
    });

    it('renders current session section with an unverified session', async () => {
        mockClient.getDevices.mockResolvedValue({ devices: [alicesDevice, alicesMobileDevice] });
        const { getByTestId } = render(getComponent());

        await act(async () => {
            await flushPromisesWithFakeTimers();
        });

        expect(getByTestId('current-session-section')).toMatchSnapshot();
    });

    it('renders current session section with a verified session', async () => {
        mockClient.getDevices.mockResolvedValue({ devices: [alicesDevice, alicesMobileDevice] });
        mockClient.getStoredDevice.mockImplementation(() => new DeviceInfo(alicesDevice.device_id));
        mockCrossSigningInfo.checkDeviceTrust
            .mockReturnValue(new DeviceTrustLevel(true, true, false, false));

        const { getByTestId } = render(getComponent());

        await act(async () => {
            await flushPromisesWithFakeTimers();
        });

        expect(getByTestId('current-session-section')).toMatchSnapshot();
    });

    it('does not render other sessions section when user has only one device', async () => {
        mockClient.getDevices.mockResolvedValue({ devices: [alicesDevice] });
        const { queryByTestId } = render(getComponent());

        await act(async () => {
            await flushPromisesWithFakeTimers();
        });

        expect(queryByTestId('other-sessions-section')).toBeFalsy();
    });

    it('renders other sessions section when user has more than one device', async () => {
        mockClient.getDevices.mockResolvedValue({
            devices: [alicesDevice, alicesOlderMobileDevice, alicesMobileDevice],
        });
        const { getByTestId } = render(getComponent());

        await act(async () => {
            await flushPromisesWithFakeTimers();
        });

        expect(getByTestId('other-sessions-section')).toBeTruthy();
    });

    describe('device detail expansion', () => {
        it('renders no devices expanded by default', async () => {
            mockClient.getDevices.mockResolvedValue({
                devices: [alicesDevice, alicesOlderMobileDevice, alicesMobileDevice],
            });
            const { getByTestId } = render(getComponent());

            await act(async () => {
                await flushPromisesWithFakeTimers();
            });

            const otherSessionsSection = getByTestId('other-sessions-section');

            // no expanded device details
            expect(otherSessionsSection.getElementsByClassName('mx_DeviceDetails').length).toBeFalsy();
        });

        it('toggles device expansion on click', async () => {
            mockClient.getDevices.mockResolvedValue({
                devices: [alicesDevice, alicesOlderMobileDevice, alicesMobileDevice],
            });
            const { getByTestId, queryByTestId } = render(getComponent());

            await act(async () => {
                await flushPromisesWithFakeTimers();
            });

            act(() => {
                const tile = getByTestId(`device-tile-${alicesOlderMobileDevice.device_id}`);
                const toggle = tile.querySelector('[aria-label="Toggle device details"]');
                fireEvent.click(toggle);
            });

            // device details are expanded
            expect(getByTestId(`device-detail-${alicesOlderMobileDevice.device_id}`)).toBeTruthy();

            act(() => {
                const tile = getByTestId(`device-tile-${alicesMobileDevice.device_id}`);
                const toggle = tile.querySelector('[aria-label="Toggle device details"]');
                fireEvent.click(toggle);
            });

            // both device details are expanded
            expect(getByTestId(`device-detail-${alicesOlderMobileDevice.device_id}`)).toBeTruthy();
            expect(getByTestId(`device-detail-${alicesMobileDevice.device_id}`)).toBeTruthy();

            act(() => {
                const tile = getByTestId(`device-tile-${alicesMobileDevice.device_id}`);
                const toggle = tile.querySelector('[aria-label="Toggle device details"]');
                fireEvent.click(toggle);
            });

            // alicesMobileDevice was toggled off
            expect(queryByTestId(`device-detail-${alicesMobileDevice.device_id}`)).toBeFalsy();
            // alicesOlderMobileDevice stayed open
            expect(getByTestId(`device-detail-${alicesOlderMobileDevice.device_id}`)).toBeTruthy();
        });
    });
});
