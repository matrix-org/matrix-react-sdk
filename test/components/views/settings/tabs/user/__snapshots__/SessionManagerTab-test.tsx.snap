// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SessionManagerTab /> renders current session section with a verified session 1`] = `
<div
  class="mx_SettingsSubsection"
  data-testid="current-session-section"
>
  <h3
    class="mx_Heading_h3 mx_SettingsSubsection_heading"
  >
    Current session
  </h3>
  <div
    class="mx_SettingsSubsection_content"
  >
    <div
      class="mx_DeviceTile"
      data-testid="device-tile-alices_device"
    >
      <div
        class="mx_DeviceTile_info"
      >
        <h4
          class="mx_Heading_h4"
        >
          alices_device
        </h4>
        <div
          class="mx_DeviceTile_metadata"
        >
          <span
            data-testid="device-metadata-isVerified"
          >
            Verified
          </span>
           · 
           · 
        </div>
      </div>
      <div
        class="mx_DeviceTile_actions"
      >
        <div
          aria-label="Toggle device details"
          class="mx_AccessibleButton mx_DeviceExpandDetailsButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_icon"
          data-testid="current-session-toggle-details"
          role="button"
          tabindex="0"
        >
          <div
            class="mx_DeviceExpandDetailsButton_icon"
          />
        </div>
      </div>
    </div>
    <br />
    <div
      class="mx_DeviceSecurityCard"
    >
      <div
        class="mx_DeviceSecurityCard_icon Verified"
      >
        <div
          height="16"
          width="16"
        />
      </div>
      <div
        class="mx_DeviceSecurityCard_content"
      >
        <p
          class="mx_DeviceSecurityCard_heading"
        >
          Verified session
        </p>
        <p
          class="mx_DeviceSecurityCard_description"
        >
          This session is ready for secure messaging.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`<SessionManagerTab /> renders current session section with an unverified session 1`] = `
<div
  class="mx_SettingsSubsection"
  data-testid="current-session-section"
>
  <h3
    class="mx_Heading_h3 mx_SettingsSubsection_heading"
  >
    Current session
  </h3>
  <div
    class="mx_SettingsSubsection_content"
  >
    <div
      class="mx_DeviceTile"
      data-testid="device-tile-alices_device"
    >
      <div
        class="mx_DeviceTile_info"
      >
        <h4
          class="mx_Heading_h4"
        >
          alices_device
        </h4>
        <div
          class="mx_DeviceTile_metadata"
        >
          <span
            data-testid="device-metadata-isVerified"
          >
            Unverified
          </span>
           · 
           · 
        </div>
      </div>
      <div
        class="mx_DeviceTile_actions"
      >
        <div
          aria-label="Toggle device details"
          class="mx_AccessibleButton mx_DeviceExpandDetailsButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_icon"
          data-testid="current-session-toggle-details"
          role="button"
          tabindex="0"
        >
          <div
            class="mx_DeviceExpandDetailsButton_icon"
          />
        </div>
      </div>
    </div>
    <br />
    <div
      class="mx_DeviceSecurityCard"
    >
      <div
        class="mx_DeviceSecurityCard_icon Unverified"
      >
        <div
          height="16"
          width="16"
        />
      </div>
      <div
        class="mx_DeviceSecurityCard_content"
      >
        <p
          class="mx_DeviceSecurityCard_heading"
        >
          Unverified session
        </p>
        <p
          class="mx_DeviceSecurityCard_description"
        >
          Verify or sign out from this session for best security and reliability.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`<SessionManagerTab /> sets device verification status correctly 1`] = `
<div
  class="mx_DeviceTile"
  data-testid="device-tile-alices_device"
>
  <div
    class="mx_DeviceTile_info"
  >
    <h4
      class="mx_Heading_h4"
    >
      alices_device
    </h4>
    <div
      class="mx_DeviceTile_metadata"
    >
      <span
        data-testid="device-metadata-isVerified"
      >
        Verified
      </span>
       · 
       · 
    </div>
  </div>
  <div
    class="mx_DeviceTile_actions"
  >
    <div
      aria-label="Toggle device details"
      class="mx_AccessibleButton mx_DeviceExpandDetailsButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_icon"
      data-testid="current-session-toggle-details"
      role="button"
      tabindex="0"
    >
      <div
        class="mx_DeviceExpandDetailsButton_icon"
      />
    </div>
  </div>
</div>
`;
